File: src/services/okfService.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";
import { isIgnoredPath, relativeToWiki } from "../fs/paths.js";
import { validateBundle, validateConceptFile } from "../okf/validate.js";
import type {
  BundleValidationReport,
  ValidationIssue,
} from "../okf/validate.js";
import type { Config } from "../config.js";
import type { SettingsService } from "./settingsService.js";

export class OkfService {
  constructor(
    private readonly config: Config,
    private readonly settings?: SettingsService,
  ) {}

  validateSingle(raw: string, relPath = "concept.md"): ValidationIssue[] {
    return validateConceptFile(relPath, raw);
  }

  async validateWikiBundle(): Promise<BundleValidationReport> {
    const files = this.collectMarkdown();
    return validateBundle(files);
  }

  collectMarkdown(): Array<{ relPath: string; content: string }> {
    const out: Array<{ relPath: string; content: string }> = [];
    const visit = (dir: string) => {
      let entries;
      try {
        entries = readdirSync(dir, { withFileTypes: true });
      } catch {
        return;
      }
      for (const entry of entries) {
        const full = path.join(dir, entry.name);
        const rel = relativeToWiki(this.config.WIKI_ROOT, full);
        if (isIgnoredPath(rel)) continue;
        if (entry.isDirectory()) visit(full);
        else if (entry.isFile() && entry.name.endsWith(".md")) {
          try {
            out.push({ relPath: rel, content: readFileSync(full, "utf8") });
          } catch {
            // unreadable file: skip
          }
        }
      }
    };
    visit(this.config.WIKI_ROOT);
    return out;
  }

  async layoutProfile(): Promise<"okf" | "wikillm"> {
    const configured = ((await this.settings?.get<string>("layout")) ??
      this.config.LAYOUT) as "auto" | "okf" | "wikillm";
    if (configured !== "auto") return configured;
    const rootIndex = path.join(this.config.WIKI_ROOT, "index.md");
    if (exists(rootIndex)) {
      const content = readFileSync(rootIndex, "utf8");
      if (content.includes("okf_version")) return "okf";
    }
    return exists(path.join(this.config.WIKI_ROOT, "wiki")) ? "wikillm" : "okf";
  }
}

function exists(p: string): boolean {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

File: rust/src/services/okf_service.rs
//! Bundle-level OKF validation over the wiki root.

use crate::config::Config;
use crate::error::Result;
use crate::okf::validate::{validate_bundle, BundleValidationReport};
use std::path::Path;

pub struct OkfService {
    config: Config,
    layout_setting: std::sync::Arc<tokio::sync::RwLock<String>>,
}

impl OkfService {
    pub fn new(config: Config, layout_setting: std::sync::Arc<tokio::sync::RwLock<String>>) -> Self {
        Self { config, layout_setting }
    }

    fn collect_markdown(root: &str) -> Vec<(String, String)> {
        fn visit(dir: &Path, root: &Path, out: &mut Vec<(String, String)>) {
            let entries = match std::fs::read_dir(dir) {
                Ok(e) => e,
                Err(_) => return,
            };
            for entry in entries.flatten() {
                let full = entry.path();
                let rel = full.strip_prefix(root).unwrap_or(&full).to_string_lossy().replace('\\', "/");
                if crate::fs::paths::is_ignored_path(&rel) {
                    continue;
                }
                if full.is_dir() {
                    visit(&full, root, out);
                } else if full.extension().map(|e| e == "md").unwrap_or(false) {
                    if let Ok(content) = std::fs::read_to_string(&full) {
                        out.push((rel, content));
                    }
                }
            }
        }
        let mut out = Vec::new();
        visit(Path::new(root), Path::new(root), &mut out);
        out
    }

    pub async fn validate_bundle(&self) -> Result<BundleValidationReport> {
        let files = Self::collect_markdown(&self.config.wiki_root);
        Ok(validate_bundle(files.iter().map(|(r, c)| (r.as_str(), c.as_str()))))
    }

    pub async fn layout_profile(&self) -> &'static str {
        let configured = self.layout_setting.read().await.clone();
        if configured != "auto" {
            return if configured == "wikillm" { "wikillm" } else { "okf" };
        }
        let root = Path::new(&self.config.wiki_root);
        let root_index = root.join("index.md");
        if root_index.is_file() {
            if let Ok(content) = std::fs::read_to_string(&root_index) {
                if content.contains("okf_version") {
                    return "okf";
                }
            }
        }
        if root.join("wiki").is_dir() { "wikillm" } else { "okf" }
    }
}
